### ISSUE TITLE
Understanding cause of bad performance

### ISSUE THREAD

**Issue by cgel** [tensorflow/tensorflow#2441]
I am training a convnet with multilple gpus and was using the cifar10 model as an example. It computes the gradients in every tower, stitches them and averages them. But that is mathematically equivalent to defining `total_loss = tf.add_n(tower_loss_list)` so I thought I would just do that.
But this implementation with 2 gpus runs at 1/2 the speed of the single gpu implementation. I am guessing the reason is that the gradient ops placement is very bad, and forces a lot data to be moved around.
Do you also think that is the reason?
If so, what could be done to improve the automatic device assignment?

**Comment by mrry** [comment]
Did you change from having one `Optimizer` per tower to a single `Optimizer`? In that case, it's possible that all of the gradients are being computed on a single GPU (most likely `/gpu:0`), because the `colocate_gradients_with_ops` argument to `Optimizer.minimize()`/`Optimizer.compute_gradients()` defaults to `False`, and so the gradient ops will default to running on a single device.  You could try switching that option to `True` to see if it makes a difference.

**Comment by cgel** [comment]
Thanks for that! Yes I switched to having a single `Optimizer`.
Setting `colocare_gradients_with_ops` to true improved the performance by 50%. But still, 2 GPUs run at 3/4 the speed of 1.

**Comment by girving** [comment]
If you `tf.add_n` the losses together, you fuse the two different backward passes into a single backward pass which can no longer be parallelized across multiple GPUs (TensorFlow currently does not split single ops across multiple GPUs).  Thus, one of the GPUs is likely running idle during the backward pass.

**Comment by cgel** [comment]
That is not convincing. Of course the gradient of ```tf.add_n()``` runs as a single op, but the backpropagation should continue independently throughout the nodes in each separate GPU.
This is evidenced by the fact that all GPUs run at the same level -- 30% or so.

**Comment by girving** [comment]
@cgel: Apologies, for the brief hallucination: what I said above is of course wrong.

**Comment by girving** [comment]
@vrv: Do you have any guesses as to the problem?

**Comment by vrv** [comment]
No guesses, it would be useful to use the timeline.py and chrome trace viewer to see where there may be gaps in execution.

**Comment by vrv** [comment]
Without more information there's not much we can do -- closing for now but feel free to reopen if you have more information!
